pkg/libartifact/store/store.go

```diff
@@ -177,10 +186,48 @@ func (as ArtifactStore) Add(ctx context.Context, dest string, paths []string, op
 
 	// Check if artifact exists; in GetByName not getting an
 	// error means it exists
-	if _, _, err := artifacts.GetByNameOrDigest(dest); err == nil {
+	artifact, _, err := artifacts.GetByNameOrDigest(dest)
+	if err == nil && !options.Append {
 		return nil, fmt.Errorf("artifact %s already exists", dest)
 	}
 
+	artifactManifest := specV1.Manifest{
+		Versioned:    specs.Versioned{SchemaVersion: SchemaVersion},
+		MediaType:    specV1.MediaTypeImageManifest,
+		ArtifactType: options.ArtifactType,
+		// TODO This should probably be configurable once the CLI is capable
+		Config: specV1.DescriptorEmptyJSON,
+		Layers: make([]specV1.Descriptor, 0),
+	}
+	var instanceDigest *digest.Digest
+	fileNames := map[string]struct{}{}
+	if artifact != nil && options.Append {
+		artifactManifest = artifact.Manifest.Manifest
+
+		index, err := as.readIndex()
+		if err != nil {
+			return nil, err
+		}
+		for _, manifest := range index.Manifests {
+			if manifest.Annotations[specV1.AnnotationRefName] == artifact.Name {
+				instanceDigest = &manifest.Digest
+				break
+			}
+		}
```

Comment:
Do we need to read the index? AFAICT `artifact.GetDigest()` should return the same digest.

That said it seems GetDigest() reads the full manifest again and calculates the digest so using the index like you did might actually be the best thing here

Reply:
Yes, you're right when I tried to explain why I'm doing it there. I've found that my approach is not a good one.